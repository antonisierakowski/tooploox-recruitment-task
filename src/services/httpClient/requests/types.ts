export interface SearchUserResponse {
  login: string;
  id: number;
  avatar_url: string;
  html_url: string;
  bio: string;
}

export interface SearchUserRepositoriesResponse {
  total_count: number;
  items: {
    id: number;
    name: string;
    full_name: string;
    html_url: string;
    stargazers_count: number;
  }[];
}
