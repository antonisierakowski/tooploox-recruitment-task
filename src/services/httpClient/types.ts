export enum StatusCode {
  OK = 200,
  BAD_REQUEST = 400,
  FORBIDDEN = 403,
  RESOURCE_NOT_FOUND = 404,
  UNPROCESSABLE_ENTITY = 422,
  INTERNAL_ERROR = 500,
}
