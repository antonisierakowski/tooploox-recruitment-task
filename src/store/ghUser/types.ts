export interface GhUserState {
  isLoading: boolean;
  user: GhUser;
  repositories: GhRepository[];
}

export interface GhUser {
  userName: string;
  bio: string;
  avatarUrl: string;
  url: string;
}

export interface GhRepository {
  id: number;
  name: string;
  stars: number;
  url: string;
}
