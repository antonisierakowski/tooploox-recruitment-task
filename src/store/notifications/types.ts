export interface NotificationsState {
  notifications: Notification[];
}

export interface Notification {
  notificationType: NotificationType;
  textContent: string;
  id: string;
}

export enum NotificationType {
  SUCCESS = 'success',
  FAILURE = 'failure',
}
