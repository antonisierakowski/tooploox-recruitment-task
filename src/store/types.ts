import { GhUserState } from './ghUser/types';
import { NotificationsState } from './notifications/types';

export interface RootState {
  // ghUser: GhUserState;
  // notifications: NotificationsState
}

export interface Action<TPayload = any> {
  type: string;
  payload?: TPayload;
}
