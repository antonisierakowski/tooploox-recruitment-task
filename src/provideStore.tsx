import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';

export const provideStore = (
  component: React.ReactElement,
  store: Store,
): React.ReactElement => <Provider store={store} children={component} />;
