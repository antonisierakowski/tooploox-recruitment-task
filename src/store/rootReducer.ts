import { combineReducers } from 'redux';
import { notificationsReducer } from './notifications/reducer';
import { ghUserReducer } from './ghUser/reducer';

export const rootReducer = combineReducers({
  ghUserState: ghUserReducer,
  notificationsState: notificationsReducer,
});
