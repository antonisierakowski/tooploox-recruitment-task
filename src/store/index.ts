import { createStore, applyMiddleware, Store } from 'redux';
import { rootReducer } from './rootReducer';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import { Action, RootState } from './types';
import { rootSaga } from './rootSaga';

const sagaMiddleware = createSagaMiddleware();

const middleware = composeWithDevTools(applyMiddleware(sagaMiddleware));

export const store: Store<RootState, Action> = createStore(
  rootReducer,
  middleware,
);

export const createPreloadedStore = (
  state: RootState,
): Store<RootState, Action> => createStore(rootReducer, state, middleware);

sagaMiddleware.run(rootSaga);
