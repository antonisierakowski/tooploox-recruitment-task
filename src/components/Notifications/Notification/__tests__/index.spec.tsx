import { createPreloadedStore } from '../../../../store';
import { rootStateFixture } from '../../../../fixtures/rootStateFixture';
import { render } from '@testing-library/react';
import { provideStore } from '../../../../provideStore';
import {
  onlyFailureNotificationStateFixture,
  onlySuccessNotificationStateFixture,
  onlyUnexpectedTypeNotificationStateFixture,
} from '../../../../fixtures/notifiationStateFixture';
import React from 'react';
import {
  NotificationsState,
  NotificationType,
} from '../../../../store/notifications/types';
import { RootState } from '../../../../store/types';
import { Notification } from '../index';

const store = createPreloadedStore(rootStateFixture);

const testNotifications = rootStateFixture.notificationsState.notifications;
const testIds = testNotifications.map(notification => notification.id);

it('should match to the generated snapshot', () => {
  const { asFragment } = render(
    provideStore(<Notification id={testIds[0]} />, store),
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should display the right notification given specific id', () => {
  const notificationIndex = 1;
  const expectedNotificationText =
    testNotifications[notificationIndex].textContent;

  const { container } = render(
    provideStore(<Notification id={testIds[notificationIndex]} />, store),
  );
  expect(container.textContent).toEqual(expectedNotificationText);
});

describe('getIconType helper function', () => {
  it('should return Done icon for notification type success', () => {
    const store = createPreloadedStore(
      mapNotificationsStateToRootState(onlySuccessNotificationStateFixture),
    );
    const {
      notificationType,
      id,
    } = onlySuccessNotificationStateFixture.notifications[0];
    const { getByTestId, container } = render(
      provideStore(<Notification id={id} />, store),
    );

    expect(notificationType).toBe(NotificationType.SUCCESS);
    expect(container.querySelectorAll('.notification')).toHaveLength(1);
    expect(getByTestId('successIcon')).toBeInTheDocument();
  });
  it('should return Error icon for notification type failure', () => {
    const store = createPreloadedStore(
      mapNotificationsStateToRootState(onlyFailureNotificationStateFixture),
    );
    const {
      notificationType,
      id,
    } = onlyFailureNotificationStateFixture.notifications[0];
    const { getByTestId, container } = render(
      provideStore(<Notification id={id} />, store),
    );

    expect(notificationType).toBe(NotificationType.FAILURE);
    expect(container.querySelectorAll('.notification')).toHaveLength(1);
    expect(getByTestId('failureIcon')).toBeInTheDocument();
  });
  it('should return Info icon for unexpected notification type', () => {
    const store = createPreloadedStore(
      mapNotificationsStateToRootState(
        onlyUnexpectedTypeNotificationStateFixture,
      ),
    );
    const { id } = onlyUnexpectedTypeNotificationStateFixture.notifications[0];
    const { getByTestId, container } = render(
      provideStore(<Notification id={id} />, store),
    );

    expect(container.querySelectorAll('.notification')).toHaveLength(1);
    expect(getByTestId('fallbackIcon')).toBeInTheDocument();
  });
});

const mapNotificationsStateToRootState = (
  domainState: NotificationsState,
): Partial<RootState> => {
  return {
    notificationsState: domainState,
  };
};
