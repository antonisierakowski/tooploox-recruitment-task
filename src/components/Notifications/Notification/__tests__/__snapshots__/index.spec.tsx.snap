// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match to the generated snapshot 1`] = `
<DocumentFragment>
  <div
    class="notification"
    data-testid="1"
  >
    <div
      class="notificationBody"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        data-testid="successIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
        />
      </svg>
      <p
        class="notificationText"
      >
        test1
      </p>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root closeIcon"
      data-testid="closeIcon_1"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
